here-naksha-storage-http/src/main/java/com/here/naksha/storage/http/POpToQuery.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (C) 2017-2023 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+package com.here.naksha.storage.http;
+
+import static com.here.naksha.lib.core.exceptions.UncheckedException.unchecked;
+import static com.here.naksha.lib.core.models.storage.POpType.*;
+
+import com.here.naksha.lib.core.models.storage.POp;
+import com.here.naksha.lib.core.models.storage.POpType;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.ArrayUtils;
+import org.jetbrains.annotations.NotNull;
+
+public class POpToQuery {
+
+  public static final String EQ_OPERATOR = "=";
+  public static final String NOT_EQ_OPERATOR = "!=";
+  public static final String NULL = ".null";
+  public static final String AND_DELIMITER = "&";
+
+  public static final String OR_DELIMITER = ",";
+  public static final String PATH_SEGMENT_DELIMITER = ".";
+
+  static String p0pToQuery(POp pOp) {
+    if (pOp.op() == AND) return and(pOp);
+    else return pOpToMultiValueComparison(pOp).resolve();
+  }
+
+  private static String and(POp pOp) {
+    assertHasAtLeastOneChildren(pOp);
+    return pOp.children().stream().map(POpToQuery::p0pToQuery).collect(Collectors.joining(AND_DELIMITER));
+  }
+
+  //
+  // MultiValueComparison operations
+  //
+  private static MultiValueComparison pOpToMultiValueComparison(POp pOp) {
+    if (pOp.op() == AND) throw unsupportedOperation("AND can be only a top level operation");
+    if (pOp.op() == OR) return or(pOp);
+    if (pOp.op() == NOT) return not(pOp);
+    if (pOp.op() == EXISTS) return exists(pOp);
+    return simpleLeafOperator(pOp);
+  }
+
+  private static MultiValueComparison or(POp pOp) {
+    assertHasAtLeastOneChildren(pOp);
+
+    return pOp.children().stream()
+        .map(POpToQuery::pOpToMultiValueComparison)
+        .reduce((l, r) -> {
+          if (!Objects.equals(l.operator, r.operator))
+            throw unsupportedOperation(
+                "Operators " + l.operator + " and " + r.operator + " combined in one OR");
+          if (!Objects.equals(l.path, r.path)) throw unsupportedOperation("Paths in OR are not equal");
+          return new MultiValueComparison(l.operator, l.path, ArrayUtils.addAll(l.values, r.values));
+        })
+        .get();
```

Comment:
IJ is giving me this warning: 
![image](https://github.com/xeus2001/xyz-hub/assets/9114956/3fb18a49-d252-4048-9cc4-244eb2d56da8)

If you're certain that `.get()` is safe here and value is always present, then I would apply `orElseThrow(() -> someException)` (it will be a bit safer+ IJ won't complain 😛 )

Reply:
Used new AssertionError("Should not reach here.")

Reply:
After discussion changed to IllegalStateException since Errors are to meant to be caught